[CI] IpPrefixAggregatorTests testMinDocCount failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/p7lvx4xmeq3fq/tests/:server:test/org.elasticsearch.search.aggregations.bucket.prefix.IpPrefixAggregatorTests/testMinDocCount

**Reproduction line:**
`./gradlew ':server:test' --tests "org.elasticsearch.search.aggregations.bucket.prefix.IpPrefixAggregatorTests.testMinDocCount" -Dtests.seed=27A633880C7968D3 -Dtests.locale=ar-IQ -Dtests.timezone=Africa/Maseru -Druntime.java=17`

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.search.aggregations.bucket.prefix.IpPrefixAggregatorTests&tests.test=testMinDocCount

**Failure excerpt:**
```
java.lang.AssertionError: expected:<[3]> but was:<[4]>

  at __randomizedtesting.SeedInfo.seed([27A633880C7968D3:F38AC2ED53A6D213]:0)
  at org.junit.Assert.fail(Assert.java:88)
  at org.junit.Assert.failNotEquals(Assert.java:834)
  at org.junit.Assert.assertEquals(Assert.java:118)
  at org.junit.Assert.assertEquals(Assert.java:144)
  at org.elasticsearch.search.aggregations.bucket.prefix.IpPrefixAggregatorTests.lambda$testMinDocCount$49(IpPrefixAggregatorTests.java:981)
  at org.elasticsearch.search.aggregations.AggregatorTestCase.testCase(AggregatorTestCase.java:676)
  at org.elasticsearch.search.aggregations.bucket.prefix.IpPrefixAggregatorTests.testMinDocCount(IpPrefixAggregatorTests.java:953)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

--- a/docs/changelog/89734.yaml
+++ b/docs/changelog/89734.yaml
@@ -1,0 +1,6 @@
+pr: 89734
+summary: Fix partial reduce bug in `ip_prefix`
+area: Aggregations
+type: bug
+issues:
+ - 89686

--- a/server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/InternalIpPrefix.java
+++ b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/prefix/InternalIpPrefix.java
@@ -251,7 +251,7 @@ public class InternalIpPrefix extends InternalMultiBucketAggregation<InternalIpP
                 final IteratorAndCurrent<Bucket> top = pq.top();
                 if (top.current().key.equals(value) == false) {
                     final Bucket reduced = reduceBucket(currentBuckets, reduceContext);
-                    if (reduced.getDocCount() >= minDocCount) {
+                    if (false == reduceContext.isFinalReduce() || reduced.getDocCount() >= minDocCount) {
                         reducedBuckets.add(reduced);
                     }
                     currentBuckets.clear();
@@ -272,7 +272,7 @@ public class InternalIpPrefix extends InternalMultiBucketAggregation<InternalIpP
 
             if (currentBuckets.isEmpty() == false) {
                 final Bucket reduced = reduceBucket(currentBuckets, reduceContext);
-                if (reduced.getDocCount() >= minDocCount) {
+                if (false == reduceContext.isFinalReduce() || reduced.getDocCount() >= minDocCount) {
                     reducedBuckets.add(reduced);
                 }
             }

--- a/server/src/test/java/org/elasticsearch/search/aggregations/bucket/prefix/InternalIpPrefixTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/bucket/prefix/InternalIpPrefixTests.java
@@ -1,0 +1,181 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.aggregations.bucket.prefix;
+
+import org.apache.lucene.document.InetAddressPoint;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.network.InetAddresses;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.InternalAggregations;
+import org.elasticsearch.search.aggregations.ParsedMultiBucketAggregation;
+import org.elasticsearch.test.InternalMultiBucketAggregationTestCase;
+import org.elasticsearch.test.MapMatcher;
+
+import java.net.Inet6Address;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+import static java.util.stream.Collectors.toMap;
+import static org.elasticsearch.test.MapMatcher.assertMap;
+import static org.elasticsearch.test.MapMatcher.matchesMap;
+import static org.hamcrest.Matchers.equalTo;
+
+public class InternalIpPrefixTests extends InternalMultiBucketAggregationTestCase<InternalIpPrefix> {
+    @Override
+    protected InternalIpPrefix createTestInstance(String name, Map<String, Object> metadata, InternalAggregations aggregations) {
+        return createTestInstance(name, metadata, aggregations, randomPrefixLength(), randomMinDocCount());
+    }
+
+    @Override
+    protected Class<? extends ParsedMultiBucketAggregation<?>> implementationClass() {
+        // Deprecated high level rest client not supported
+        return null;
+    }
+
+    private int randomPrefixLength() {
+        return between(1, InetAddressPoint.BYTES * 8);
+    }
+
+    private long randomMinDocCount() {
+        return randomBoolean() ? 1 : randomLongBetween(1, Long.MAX_VALUE / (maxNumberOfBuckets() + 1));
+    }
+
+    private InternalIpPrefix createTestInstance(
+        String name,
+        Map<String, Object> metadata,
+        InternalAggregations aggregations,
+        int prefixLength,
+        long minDocCount
+    ) {
+        boolean keyed = randomBoolean();
+        boolean appendPrefixLength = randomBoolean();
+        boolean canBeV4 = prefixLength <= 32;
+
+        int bucketsCount = between(1, maxNumberOfBuckets());
+        Set<BytesRef> keys = new TreeSet<>();
+        while (keys.size() < bucketsCount) {
+            boolean v4 = canBeV4 && randomBoolean();
+            byte[] ip = InetAddressPoint.encode(randomIp(v4));
+            byte[] mask = mask(v4 ? prefixLength + 96 : prefixLength);
+            byte[] subnet = new byte[InetAddressPoint.BYTES];
+            for (int i = 0; i < InetAddressPoint.BYTES; i++) {
+                subnet[i] = (byte) (ip[i] & mask[i]);
+            }
+            keys.add(new BytesRef(ip));
+        }
+        List<InternalIpPrefix.Bucket> buckets = new ArrayList<>(keys.size());
+        for (Iterator<BytesRef> itr = keys.iterator(); itr.hasNext();) {
+            BytesRef key = itr.next();
+            boolean v6 = InetAddressPoint.decode(key.bytes) instanceof Inet6Address;
+            buckets.add(
+                new InternalIpPrefix.Bucket(
+                    DocValueFormat.IP,
+                    key,
+                    keyed,
+                    v6,
+                    prefixLength,
+                    appendPrefixLength,
+                    randomLongBetween(0, Long.MAX_VALUE),
+                    aggregations
+                )
+            );
+        }
+
+        return new InternalIpPrefix(name, DocValueFormat.IP, keyed, minDocCount, buckets, metadata);
+    }
+
+    private byte[] mask(int prefixLength) {
+        byte[] mask = new byte[InetAddressPoint.BYTES];
+        int m = 0;
+        int b = 0x80;
+        for (int i = 0; i < prefixLength; i++) {
+            mask[m] |= b;
+            b = b >> 1;
+            if (b == 0) {
+                m++;
+                b = 0x80;
+            }
+        }
+        return mask;
+    }
+
+    @Override
+    protected BuilderAndToReduce<InternalIpPrefix> randomResultsToReduce(String name, int size) {
+        Map<String, Object> metadata = createTestMetadata();
+        InternalAggregations aggregations = createSubAggregations();
+        int prefixLength = randomPrefixLength();
+        long minDocCount = randomMinDocCount();
+        List<InternalIpPrefix> inputs = new ArrayList<>(size);
+        for (int i = 0; i < size; i++) {
+            InternalIpPrefix t = createTestInstance(name, metadata, aggregations, prefixLength, minDocCount);
+            inputs.add(t);
+        }
+        return new BuilderAndToReduce<>(mockBuilder(inputs), inputs);
+    }
+
+    @Override
+    protected void assertReduced(InternalIpPrefix reduced, List<InternalIpPrefix> inputs) {
+        InternalIpPrefix leader = inputs.get(0);
+        assertThat(reduced.keyed, equalTo(leader.keyed));
+        assertThat(reduced.format, equalTo(leader.format));
+        assertThat(reduced.minDocCount, equalTo(leader.minDocCount));
+        Map<BytesRef, Long> expectedCounts = new HashMap<>();
+        for (InternalIpPrefix i : inputs) {
+            for (InternalIpPrefix.Bucket b : i.getBuckets()) {
+                assertThat(b.getFormat(), equalTo(DocValueFormat.IP));
+                long acc = expectedCounts.getOrDefault(b.getKey(), 0L);
+                acc += b.getDocCount();
+                expectedCounts.put(b.getKey(), acc);
+            }
+        }
+        MapMatcher countsMatches = matchesMap();
+        for (Map.Entry<BytesRef, Long> e : expectedCounts.entrySet()) {
+            if (e.getValue() >= leader.minDocCount) {
+                countsMatches = countsMatches.entry(DocValueFormat.IP.format(e.getKey()), e.getValue());
+            }
+        }
+        assertMap(
+            new TreeMap<>(reduced.getBuckets().stream().collect(toMap(b -> b.getKeyAsString(), b -> b.getDocCount()))),
+            countsMatches
+        );
+    }
+
+    public void testPartialReduceNoMinDocCount() {
+        InternalIpPrefix.Bucket b1 = new InternalIpPrefix.Bucket(
+            DocValueFormat.IP,
+            new BytesRef(InetAddressPoint.encode(InetAddresses.forString("192.168.0.1"))),
+            false,
+            false,
+            1,
+            false,
+            1,
+            InternalAggregations.EMPTY
+        );
+        InternalIpPrefix.Bucket b2 = new InternalIpPrefix.Bucket(
+            DocValueFormat.IP,
+            new BytesRef(InetAddressPoint.encode(InetAddresses.forString("200.0.0.1"))),
+            false,
+            false,
+            1,
+            false,
+            2,
+            InternalAggregations.EMPTY
+        );
+        InternalIpPrefix t = new InternalIpPrefix("test", DocValueFormat.IP, false, 100, List.of(b1, b2), null);
+        InternalIpPrefix reduced = (InternalIpPrefix) t.reduce(List.of(t), emptyReduceContextBuilder().forPartialReduction());
+        assertThat(reduced.getBuckets().get(0).getDocCount(), equalTo(1L));
+        assertThat(reduced.getBuckets().get(1).getDocCount(), equalTo(2L));
+    }
+}

--- a/server/src/test/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregatorTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/bucket/prefix/IpPrefixAggregatorTests.java
@@ -41,6 +41,7 @@ import java.util.function.Function;
 import java.util.stream.Collectors;
 
 import static java.util.Collections.singleton;
+import static org.hamcrest.Matchers.equalTo;
 
 public class IpPrefixAggregatorTests extends AggregatorTestCase {
 
@@ -928,7 +929,6 @@ public class IpPrefixAggregatorTests extends AggregatorTestCase {
     }
 
     public void testMinDocCount() throws IOException {
-        // GIVEN
         final int prefixLength = 16;
         final String field = "ipv4";
         int minDocCount = 2;
@@ -949,22 +949,19 @@ public class IpPrefixAggregatorTests extends AggregatorTestCase {
             new TestIpDataHolder("10.122.2.67", "10.122.0.0", prefixLength, defaultTime())
         );
 
-        // WHEN
         testCase(aggregationBuilder, new MatchAllDocsQuery(), iw -> {
             for (TestIpDataHolder ipDataHolder : ipAddresses) {
                 iw.addDocument(
                     singleton(new SortedDocValuesField(field, new BytesRef(InetAddressPoint.encode(ipDataHolder.getIpAddress()))))
                 );
             }
-        }, agg -> {
-            final InternalIpPrefix ipPrefix = (InternalIpPrefix) agg;
+        }, (InternalIpPrefix ipPrefix) -> {
             final Set<String> expectedSubnets = Set.of("192.168.0.0");
             final Set<String> ipAddressesAsString = ipPrefix.getBuckets()
                 .stream()
                 .map(InternalIpPrefix.Bucket::getKeyAsString)
                 .collect(Collectors.toUnmodifiableSet());
 
-            // THEN
             ipPrefix.getBuckets().forEach(bucket -> {
                 assertFalse(bucket.isIpv6());
                 assertFalse(bucket.appendPrefixLength());
@@ -978,9 +975,9 @@ public class IpPrefixAggregatorTests extends AggregatorTestCase {
             assertTrue(
                 ipPrefix.getBuckets().stream().map(InternalIpPrefix.Bucket::getDocCount).allMatch(docCount -> docCount >= minDocCount)
             );
-            assertEquals(
+            assertThat(
                 ipPrefix.getBuckets().stream().sorted(IP_ADDRESS_KEY_COMPARATOR).map(InternalIpPrefix.Bucket::getDocCount).toList(),
-                List.of(4L)
+                equalTo(List.of(4L))
             );
         }, fieldType);
     }

--- a/test/framework/src/main/java/org/elasticsearch/test/InternalMultiBucketAggregationTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/InternalMultiBucketAggregationTestCase.java
@@ -99,6 +99,11 @@ public abstract class InternalMultiBucketAggregationTestCase<T extends InternalA
 
     protected abstract T createTestInstance(String name, Map<String, Object> metadata, InternalAggregations aggregations);
 
+    /**
+     * The parsed version used by the deprecated high level rest client or
+     * {@code null} if the deprecated high level rest client isn't supported
+     * by this agg.
+     */
     protected abstract Class<? extends ParsedMultiBucketAggregation<?>> implementationClass();
 
     @Override
@@ -118,6 +123,13 @@ public abstract class InternalMultiBucketAggregationTestCase<T extends InternalA
     public void testIterators() throws IOException {
         final T aggregation = createTestInstanceForXContent();
         assertMultiBucketsAggregations(aggregation, parseAndAssert(aggregation, false, false), true);
+    }
+
+    @Override
+    protected <P extends ParsedAggregation> P parseAndAssert(InternalAggregation aggregation, boolean shuffled, boolean addRandomFields)
+        throws IOException {
+        assumeFalse("deprecated high level rest client not supported", implementationClass() == null);
+        return super.parseAndAssert(aggregation, shuffled, addRandomFields);
     }
 
     @Override

